thetis/adr_model.py

```diff
@@ -0,0 +1,220 @@
+                "reaction":
+                {
+                    "expression": reaction_term,
+                    "args": list(map(str, reaction_args)),
+                    "function": lambdify(
+                        reaction_args, reaction_term,
+                        modules=lambdify_modules)
+                }
+            }
+
+    def _validate_keys(self):
+        # Check that none of the constant or species keys are reserved symbols
+        for k in (self.constant_keys | self.species_keys):
+            if k in ADR_Model.reserved_symbols:
+                print("Symbol \"%s\" is reserved." % k)
+                raise RuntimeError
+        # Check that self.constant_keys and self.species_keys do not have any
+        # elements in common
+        shared_keys = self.constant_keys & self.species_keys
+        if len(shared_keys) != 0:
+            print("The following keys are shared between a species "
+                  "and a constant: %s." % ', '.join(shared_keys))
+            raise RuntimeError
+
+    def _check_all_symbols_recognised(self, expression):
+        for s in expression.free_symbols:
+            symbol_str = str(s)
+            if symbol_str in ADR_Model.reserved_symbols:
+                continue
+            if symbol_str in self.constant_keys:
+                continue
+            if symbol_str in self.species_keys:
+                continue
+            print("Unrecognised symbol \"%s\" found in the following "
+                  "expression:" % symbol_str)
+            print(expression)
+            raise RuntimeError
+
+    def _substitute_constants(self, expression):
+        """
+        Substitutes numeric values of `self.constants` into a SymPy expression.
+
+        :arg expression: The expression into which the constant values are
+            to be substituted.
+        :type expression: SymPy expression
+        :return: The updated expression with constants replaced by numeric
+            values.
+        :rtype: SymPy expression
+        """
+        for s in expression.free_symbols:
+            constant_name = str(s)
+            if constant_name in self.constants.keys():
+                constant_value = self.constants[constant_name]
+                expression = expression.subs(
+                    sympy.symbols(constant_name),
+                    constant_value)
+        return expression
+
+    def species_name(self, k):
+        return self.species[k]["name"]
+
+    def diffusion(self, k):
+        return self.species[k]["diffusion"]
+
+    def reaction_term(self, k):
+        return self.species[k]["reaction"]["expression"]
+
+    def reaction_args(self, k):
+        return self.species[k]["reaction"]["args"]
+
+    def reaction_function(self, k):
+        return self.species[k]["reaction"]["function"]
+
+    def list_species_keys(self):
+        return sorted(
+            list(self.species_keys),
+            key=lambda k: self.species[k]['index'])
+
+    def dependency_graph(self):
```

Comment:
This method isn't used anywhere, and it's the only reason the NetworkX import is there. I've kept it in to allow for discussion in this PR, in case developers decide to make use of it somehow.

Reply:
I think it's useful functionality. Presumably pip installing networkx is pretty quick?

Reply:
Please could you use this in the new version of the Gray-Scott example?